docs: Module 1 and 2 (multiple Sections) (B9lab content updates 1) (#791)
cosmos/sdk-tutorials: b9lab-content/2-what-is-cosmos/03-cosmos-ecosystem.md

More than this, Cosmos is also **a blockchain ecosystem**, complete with protocols, SDK, tokens, wallets, applications, repositories, services, and tools.

Let's review them.

## A Whole Universe to Discover - Tokens, Wallets, Apps, and Services

Fittingly, Cosmos is an ever-expanding ecosystem of tokens, wallets, tools, and interconnected apps and services, all built for the decentralized future.

At the time of writing, almost **$100 billion in digital assets** are under management and secured by Cosmos. Digital assets on Cosmos include fungible and non-fungible tokens (NFT). You can issue in-application tokens to conduct settlements, bespoke issuance, inflation/deflation, and more.

Among the fungible tokens secured by Cosmos are the [Binance Coin](https://www.binance.com/en/bnb), [Terra](https://www.terra.money/), and of course, ATOM. Remember that because they are defined on application-specific blockchains, their developers are free from the constraints of a hypothetical general-purpose blockchain.

In addition to the vast number of tokens, a variety of applications and services, as well as wallets, and explorers are based on Cosmos.

> Currently, **249 applications and services** build on Cosmos. At the time of writing, most applications and projects deal with finance, closely followed by infrastructure. Privacy, marketplace, social impact, and social applications and projects are less common.

Reviewer: Very good idea and a good catch. Thank you.

<HighlightBox type="tip">

Do you want an up-to-date list of applications and services built on Cosmos? Head to the [Cosmos ecosystem overview](https://cosmos.network/ecosystem/apps/?ref=cosmonautsworld)!

</HighlightBox>

As you might expect, not all applications are at the same stage of development. Namely, deployment on the mainnet has been conducted for most applications and projects. The rest is currently either in proof of concept, development, or was deployed to a testnet only.

> Moreover, [**33 wallets**](https://cosmos.network/ecosystem/wallets) and block explorers for Cosmos are part of the ecosystem. Most are for Android, such as the [Atomic Walllet](https://atomicwallet.io/) and [Coinex](https://www.coinex.com/), or iOS, like [AirGap](https://airgap.it/) and [Wallet.io](https://walletio.io/), but you can also find a number of web wallets, like [Exodus](https://www.exodus.com) and [Keplr](https://wallet.keplr.app/).

Reviewer 2: ```suggestion
Moreover, [**33 wallets**](https://cosmos.network/ecosystem/wallets) and block explorers for Cosmos are part of the ecosystem. Most are for Android, such as the [Atomic Wallet](https://atomicwallet.io/) and [Coinex](https://www.coinex.com/), or iOS, like [AirGap](https://airgap.it/) and [Wallet.io](https://walletio.io/), but you can also find a number of web wallets, like [Exodus](https://www.exodus.com) and [Keplr](https://wallet.keplr.app/).
```

## Main Components

The main aim of the Cosmos network is to provide an **ecosystem for easy blockchain development** based on the Tendermint BFT and the Inter-Blockchain Communication protocol (IBC) thanks to the so-called [Cosmos SDK](https://v1.cosmos.network/sdk).

> Each chain in the Cosmos ecosystem relies on **fast-finality BFT consensus algorithms**, such as the [Tendermint](https://github.com/tendermint/tendermint) consensus. This ensures a common consensus mechanism at work in all chains of the network. Tendermint itself is used in case of the [Cosmos Hub](https://hub.cosmos.network/main/hub-overview/overview.html), [IRIS Hub](https://www.irishub.rw/), [Binance Chain](https://www.binance.org/en), [Terra](https://www.terra.money/), and [Kava](https://www.kava.io/).

Reviewer 3: Tendermint is used by every chain developing with the Cosmos-SDK, and more

> Before Cosmos came along, developing a smart contract was much easier than building a whole new chain. Now, with the Cosmos SDK, entire flexible, secure, high performance, and sovereign **application-specific blockchains** can be developed. To allow this, building modular, adaptable, and interchangeable open-source development tools is at the center of Cosmos' mission.

Reviewer 3: ```suggestion
Before Cosmos came along, developing a whole new chain was much more difficult and expensive than building a smart contract. Now, with the Cosmos SDK, entire flexible, secure, high performance, and sovereign **application-specific blockchains** can be developed. To allow this, building modular, adaptable, and interchangeable open-source development tools is at the center of Cosmos' mission.
```

The ready-built modules of the Cosmos SDK can be easily used - It is as simple as importing modules into the application. Additionally, developers can create their own modules to introduce specific functionalities. It is expected that, with the growth of the ecosystem, the number of modules will grow, thereby facilitating the development of more complex applications.

*Building on modular components, many of those you did not write yourself? Although it sounds like an increased potential for attacks, and faulty or malicious nodes operating undetected? One does not have to worry.*

The Cosmos SDK is built on a foundation of the [object-capability model](https://docs.cosmos.network/master/core/ocap.html). It not only favors modularity, it also encapsulates code implementation. Thus, capabilities constrain security boundaries between modules.

The default consensus mechanism available to build with the SDK is [Tendermint Core](https://docs.tendermint.com/master/).

## The Inter-Blockchain Communication Protocol

> The [Inter-Blockchain Communication (IBC) protocol](https://ibcprotocol.org/) is the basis for **interoperability** in Cosmos. It leverages Tendermint's instant finality to allow for the transfer of value (i.e. tokens) and data communication between heterogenous chains. Blockchains with different applications and architecture specifications become interoperable whether or not they share a validator set.

Reviewer 3: Note it also work with non-Tendermint chains, even without instant finality. The ICS 02 Client interface is generic over consensus. ICS 07 is the Tendermint implementation for this, but there are others

> Without IBC, the interoperability of heterogenous chains is difficult to achieve because they may implement the consensus, networking, and application layers in various ways. As soon as a blockchain is compatible with IBC, it becomes interoperable with others. IBC has few requirements to make a chain compatible: Mainly, the consensus layer must have fast and absolute finality.

Reviewer 3: > Mainly, the consensus layer must have fast and absolute finality

Nope we actually generalized :)

Really the requirement is that it have some form of light client.

*With these chain-to-chain connections now possible, how can we create a network of blockchains?*

The first naive idea would be to connect each blockchain with the other chains through direct IBC connections. Unfortunately, this would quickly lead to a very large, expensive, and ultimately unmanageable number of connections in the network. In fact, the number of connections in the network would increase quadratically with the number of blockchains.

A second idea, one that **reduces the number of connections necessary** back down to the number of chains, would be to introduce an intermediate entity that facilitates the transfer between chains, in a hub and spoke model. Unfortunately, this would bring back unwelcome centralization.

*How could we uphold decentralization then?*

First, by keeping the capability of chain-to-chain connections made possible by **IBC**. Second, by creating a flexible multi-hub and spoke model whereby a spoke gets automatic access to all the other spokes of the hub it connected to. This is the road **Cosmos** took.

> Cosmos implements a **modular architecture with two blockchain classes**: **hubs** and **zones**.

Reviewer 3: Should be more clear there that Hubs are just Zones with a lot of IBC connections. Some folks think the Hub/Zone terminology is confusing and should be abandoned all together. But it should be mentioned that there's no enforced topology of IBC chains, they can connect to whoever they want, but the expectation is that a hub and spoke model will emerge as some chains come to specialize in providing IBC connectivity to many others

--> Include hotspot image on zones and hubs

An example of such a specific chain is the **[Cosmos Hub](https://hub.cosmos.network/main/hub-overview/overview.html)**, the first hub to be created. It is a public, Proof-of-Stake (PoS) blockchain with a native token, ATOM. The Cosmos Hub can be understood as a router facilitating, among other things, transactions between the chains connected to it. For example, transaction fees can be paid in different tokens as long as the zone trusts the Cosmos Hub and the other zones connected to it.

*How do we connect our chain to a non-Tendermint chain?* In fact, the IBC connection is not limited to Tendermint-based chains.

If the other non-Tendermint chain is a **fast-finality chain**, i.e. a blockchain using a fast-finality consensus algorithm, the connection can be established by adapting IBC to work with the non-Tendermint consensus mechanism.

> If the other chain is a **probabilistic-finality chain**, i.e. a blockchain without fast finality like those relying on Proof-of-Work (PoW), a simple adaptation of IBC is not sufficient. Instead, a more elaborate setup is required. More precisely, a proxy-chain, a so-called **peg-zone**, helps establish interoperability. Peg-zones are fast-finality blockchains tracking chain states to establish finality. The peg-zone chain itself is IBC-compatible and acts as the **bridge** between the rest of the IBC network, i.e. its blockchains, and the probabilistic-finality chain.

Reviewer 3: Not quite true. The real limit is the ability of of the other chain to implement the core IBC standards (primarily ICS02, 03, 04).

Reviewer 3: instead of these two paragraphs about fast finality or not I would just say the only requirement is that it be possible to implement a *light client* for the chain. Then we only need to implement the ICS02 client interface for that chains light client and it can connect over IBC. of course additionally the chain will need the ability to implement the rest of IBC logic, so it might need some smart contracting facility. so bitcoin for instance, even though it has a light client, it doesnt have enough functionality to allow you to implemenet IBC on Bitcoin. But with Ethereum, you can implement IBC in solidity and do a full Cosmos<->Ethereum IBC connection. But with Avalanche, even though it's fast finality, we don't know how to do a light client, so we're not sure how to do IBC yet ... hope that helps!

Reviewer: Thank you @ebuchman! That definitely helps.

Reviewer 4: Is this proposal acceptable? 

1. It must be possible to implement a *light client* that is capable of exploring and interacting with the chain without necessarily syncing the entire chain. 
2. It must be possible to implement the [ICS02 client interface](https://github.com/cosmos/ibc/blob/master/spec/core/ics-002-client-semantics/README.md) with the light client. The ICS02 client interface usually requires smart contract capability on the other chain where important parts of the interface will be implemented. For this reason, most, but not all, blockchains with light client capability are theoretically compatible with IBC. For example, Ethereum has both light client and smart contract capability, so and IBC bridge to Ethereum is feasible.

## CosmWasm - Multi-Chain Smart Contracts

[CosmWasm](https://cosmwasm.com/) is a multi-chain platform for building robust DApps for Cosmos using smart contracts. It builds on Tendermint and the Cosmos SDK.

Its key features are:

* Mature tools for the development and testing of smart contracts,
* Close integration with the Cosmos SDK and ecosystem, and
* A secure architecture to avoid attack vectors.

With CosmWasm, smart contracts can run on multiple chains with the help of the IBC protocol. It adds further flexibility for developers and makes smart contract development faster.

CosmWasm is written as a module to be plugged into the Cosmos SDK, and leverages the speed of Wasm and the power of Rust.

<HighlightBox type="tip">

For a deeper dive, see the [CosmWasm documentation](https://docs.cosmwasm.com/docs/)

</HighlightBox>

## The Possibility of Using Alternative Blockchain Frameworks and SDKs

As the Cosmos SDK is modular, developers can port existing codebases in Go on top of the SDK. This gives developers the opportunity to build on Cosmos without having to comprise too much on the tool set and environment used.

> For example, with [Ethermint](https://github.com/cosmos/ethermint) developers can port the Ethereum Virtual Machine (EVM) to make it a Cosmos SDK module compatible and combinable with existing modules. Thus, Ethermint blockchains are compatible with Cosmos.

Reviewer 3: ```suggestion
For example, with [Ethermint](https://github.com/cosmos/ethermint) developers can utilize the Ethereum Virtual Machine (EVM) from the main Go Ethereum client as a Cosmos SDK module compatible and combinable with existing modules. Thus, Ethermint blockchains are compatible with Cosmos.
```
